[logs] Add support for new intake api

vickenty opened the pull request:
### What does this PR do?

Add support for new version 2 intake API used for logs and events. New API supports additional parameters to differentiate requests: a track type, which is part of a url, and an optional `dd-protocol`
header.

Each component that interacts that interacts with the intake, was assigned an individual type; this is passed via various routes into the endpoint construction.

Whether new API is used or not is controlled by a new hidden `use_v2_api` configuratio option (available, like other options, for each agent component separately).

Additional endpoints inherit the api version setting from the parent endpoint, but it can be overridden if necessary.

### Describe how to test your changes

Agent Core to test the logs agent (instructions below). Other teams please use a variation of this plan as applicable to your part of the agent.

Configure a logs check and enable logs agent:

- Configure a logs check and enable logs agent (using default v1 api).
  ```yaml
  logs_config:
    enabled: true
  ```
- Check in the agent log that the logs connectivity check is done to `/v1/input` and succeeds.
  ```
  [...] Sending HTTP connectivity request to https://agent-http-intake.logs.datadoghq.com/v1/input
  [...] HTTP connectivity successful
  ```
- Add some log entries to the log file and check that they appear in the app.
- Enable v2 api and restart the agent.
  ```yaml
  logs_config:
    enabled: true
    use_v2_api: true
  ```
- Check in the agent log that the logs connectivity check is done to `/api/v2/logs` and succeeds.
  ```
  [...] Sending HTTP connectivity request to https://agent-http-intake.logs.datadoghq.com/api/v2/logs
  [...] HTTP connectivity successful
  ```
- Add some log entries to the log file and check that they appear in the app.

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.

vickenty committed:
[logs] Add support for new intake api

Add support for new version 2 intake API used for logs and events.
New API supports additional parameters to differentiate requests: a
track type, which is part of a url, and an optional `dd-protocol`
header.

Each component that interacts that interacts with the intake, was
assigned an individual type; this is passed via various routes into the
endpoint construction.

Whether new API is used or not is controlled by a new hidden
`use_v2_api` configuratio option (available, like other options, for
each agent component separately).

Additional endpoints inherit the api version setting from the parent
endpoint, but it can be overridden if necessary.

joyzhang-dd commented on pkg/epforwarder/epforwarder.go:
```diff
@@ -39,6 +40,7 @@ var passthroughPipelineDescs = []passthroughPipelineDesc{
 		eventType:              eventTypeDBMMetrics,
 		endpointsConfigPrefix:  "database_monitoring.metrics.",
 		hostnameEndpointPrefix: "dbm-metrics-intake.",
+		intakeTrackType:        "dbmetrics",
```
Hey @justiniso, I know the dbmetrics has not officially onboarded onto event-platform, but looking forwards, would you confirm `dbmetrics` is the track name you will use?

justiniso replied:
Since this is not actually defined anywhere other than feature branches, I think we should switch this to `dbmmetrics` to remain consistent with our intake endpoint.

joyzhang-dd commented on cmd/security-agent/app/app.go:
```diff
@@ -44,8 +44,12 @@ import (
+	// loggerName is the name of the security agent logger
+	loggerName coreconfig.LoggerName = "SECURITY"
+	// No intake track type is defined for the security agent (yet), we use v1 intake only.
+	intakeTrackType = ""
```
track types are defined for the security agent. We should switch to v2 as well.
the things, I think the security agent are sending to two DNS:
- `compliance-http-intake.logs.` in [here](https://github.com/DataDog/datadog-agent/blob/main/cmd/security-agent/app/compliance.go#L60)  This is the track-type is `compliance`
- `runtime-security-http-intake.logs.` [here](https://github.com/DataDog/datadog-agent/blob/main/cmd/security-agent/app/runtime.go#L152)  This is the track-type is `logs`

cc @lebauce and @nicolas-renaud Could I trouble you to confirm

djova reviewed (approved):
👍   from Database Monitoring

joyzhang-dd commented on cmd/security-agent/app/app.go:
```diff
@@ -115,12 +119,12 @@ func init() {
 }
 
 func newLogContext(logsConfig *config.LogsConfigKeys, endpointPrefix string) (*config.Endpoints, *client.DestinationsContext, error) {
```
since the `newLogsContext` is called in two separate place, we probably need to pass the track-type in

joyzhang-dd commented on pkg/logs/logs.go:
```diff
@@ -68,12 +70,12 @@ func start(getAC func() *autodiscovery.AutoConfig, serverless bool, logsChan cha
+	endpoints, err := config.BuildEndpoints(httpConnectivity, intakeTrackType, intakeProtocol)
 	if serverless {
-		endpoints, err = config.BuildServerlessEndpoints()
+		endpoints, err = config.BuildServerlessEndpoints(intakeTrackType, intakeProtocol)
```
for some historical reason (which we hope to change in the future), the Intake backend considers that the json send by the agent is different than by lambda. So
- the normal HTTP endpoint: we set the intakeProtocol to `agent-json`
- the serverless Endpoint: we don't set the intakeProtocol.

vickenty committed:
[logs] use correct protocol for serverless agent

vickenty committed:
[security-agent] use correct track for intake

Security agent interacts with two different intake endpoints, and needs
to use a specific track type with each.

vickenty committed:
[epforwarder] update track type for dbm metrics

clamoriniere commented on pkg/logs/client/http/destination.go:
```diff
@@ -232,7 +237,13 @@ func buildURL(endpoint config.Endpoint) string {
 		address = endpoint.Host
 	}
-	return fmt.Sprintf("%v://%v/v1/input", scheme, address)
+	uri := ""
+	if endpoint.Version == config.EPIntakeVersion2 && endpoint.TrackType != "" {
+		uri = fmt.Sprintf("api/v2/%s", endpoint.TrackType)
+	} else {
+		uri = "v1/input"
+	}
+	return fmt.Sprintf("%v://%v/%v", scheme, address, uri)
```
Nits,
I think you can use the [net/url](https://golang.org/pkg/net/url/#URL) package here to build the URI.

vickenty replied:
Good catch, thanks!

vickenty committed:
[logs] use net/url to build the destination url
